coderd/rbac/USAGE.md

```diff
@@ -0,0 +1,344 @@
+        Actions:  []policy.Action{policy.ActionRead},
+        Resource: rbac.ResourceFrobulator.WithOwner(currentUser.String()).InOrg(orgID),
+        AuthorizeMap: map[bool][]hasAuthSubjects{
+            true:  {orgMemberMe, orgAdmin, owner, orgAuditor},
+            false: {setOtherOrg, memberMe, templateAdmin, userAdmin, orgTemplateAdmin, orgUserAdmin},
+        },
+    },
+```
+
+Note how the `FrobulatorsModify` test case is just validating the
+`policy.ActionCreate, policy.ActionUpdate, policy.ActionDelete` actions, and
+only the **orgMember**, **orgAdmin**, and **owner** can access it.
+
+Similarly, the `FrobulatorsReadOnly` test case is only validating
+`policy.ActionRead`, which is allowed on all of the above plus the
+**orgAuditor** role.
+
+Now the tests pass, because we have covered all the possible scenarios:
+
+```bash
+$ go test github.com/coder/coder/v2/coderd/rbac -count=1
+ok  	github.com/coder/coder/v2/coderd/rbac	1.313s
+```
+
+When a case is not covered, you'll see an error like this (I moved the
+`orgAuditor` option from `true` to `false):
+
+```bash
+--- FAIL: TestRolePermissions (0.79s)
+    --- FAIL: TestRolePermissions/FrobulatorsReadOnly (0.01s)
+        roles_test.go:737:
+            	Error Trace:	/tmp/coder/coderd/rbac/roles_test.go:737
+            	Error:      	An error is expected but got nil.
+            	Test:       	TestRolePermissions/FrobulatorsReadOnly
+            	Messages:   	Should fail: FrobulatorsReadOnly as "org_auditor" doing "read" on "frobulator"
+FAIL
+FAIL	github.com/coder/coder/v2/coderd/rbac	1.390s
+FAIL
+```
+
+This shows you that the `org_auditor` role has `read` permissions on the
+frobulator, but no test case covered it.
+
+**NOTE: don't just add cases which make the tests pass; consider all the way in
+which your resource must be used, and test all of those scenarios!**
+
+# Database authorization
+
+Now that we have the RBAC system fully configured, we need to make use of it.
+
+Let's add a SQL query to `coderd/database/queries/frobulators.sql`:
+
+```sql
+-- name: GetUserFrobulators :many
+SELECT *
+FROM frobulators
+WHERE user_id = @user_id::uuid;
+```
+
+Once we run `make gen`, we'll find some stubbed code in
+`coderd/database/dbauthz/dbauthz.go`.
+
+```go
+...
+func (q *querier) GetUserFrobulators(ctx context.Context) ([]database.Frobulator, error) {
+    panic("not implemented")
+}
+...
+```
+
+Let's modify this function:
+
+```go
+...
+func (q *querier) GetUserFrobulators(ctx context.Context, userID uuid.UUID) ([]database.Frobulator, error) {
+	if err := q.authorizeContext(ctx, policy.ActionRead, rbac.ResourceFrobulator.WithOwner(userID.String())); err != nil {
+		return nil, err
+	}
+	return q.db.GetUserFrobulators(ctx, userID)
+}
+...
+```
+
+This states that the `policy.ActionRead` permission is required in this query on
+the `ResourceFrobulator` resources, and `WithOwner(userID.String())` specifies
+that this user must own the resource.
```

Comment:
> `WithOwner(userID.String())` specifies that this user must own the resource

The `WithOwner` and `InOrg` specify what level of permissions can be used. If you refer back to the [readme roles table](https://github.com/coder/coder/blob/ceffff9b0676ef708e6ccf40d24279140824e670/coderd/rbac/README.md#roles) there are 3 levels of permissions that act hierarchical. 

Site will always apply, no matter what. 

If `InOrg` is omitted, then the `Org` column will always read `_` (abstain). No org permissions can apply.

If `WithOwner` is omitted, then the `user` column will always read `_`.